### PR TITLE
Add host.json setting for new Distributed Tracing feature

### PR DESCRIPTION
**Pull request by bachuv**
This PR adds a host.json setting so customers can easily configure the new distributed tracing feature. 

A new `TelemetryModule` called `DurableFunctionsTelemetryModule` from DurableTask.ApplicationInsights is used if the new host.json setting, `DurableDistributedTracingVersion`, is set to `V2`.

```
"durableTask": {
      "tracing": {
        "DistributedTracingEnabled": true,
        "Version": "V2"
      }
    }
```

Related DurableTask PR - https://github.com/Azure/durabletask/pull/871

image

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)
* [ ] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
    * [ ] Otherwise: That work is being tracked here: #issue_or_pr_in_each_sdk
* [ ] My changes **do not** change the version of the WebJobs.Extensions.DurableTask package
    * [ ] Otherwise: major or minor version updates are reflected in `/src/Worker.Extensions.DurableTask/AssemblyInfo.cs`
* [ ] My changes **do not** add EventIds to our EventSource logs
    * [ ] Otherwise: Ensure the EventIds are within the supported range in our existing Windows infrastructure. You may validate this with a deployed app's telemetry. You may also extend the range by completing a PR such as [this one](https://msazure.visualstudio.com/One/_git/AAPT-Antares-Websites/pullrequest/7463263?_a=files).

### CODE REVIEW

File: src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs
Comment (bachuv) [inline src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs:73]: I left this name for now, but I want to change it before merging since it's not very self-explanatory. The current setting `DistributedTracingEnabled` is used in various different places for the previous implementation so I was hesitant to use that setting. I'm open to suggestions on how to handle this.

File: src/WebJobs.Extensions.DurableTask/Correlation/ITelemetryActivator.cs
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/Correlation/ITelemetryActivator.cs:18]: Do we have alternative places in which to store this? Otherwise, I'd think this is fine enough :)

File: src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs:73]: Are we looking to completely replace the old implementation of Distributed Tracing once this goes into public preview? If so - I'd see the benefit in re-using the name.

Comment (bachuv) [reply src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs:73]: The plan right now is to keep the old implementation until GA.

File: src/WebJobs.Extensions.DurableTask/Correlation/ITelemetryActivator.cs
Comment (bachuv) [reply src/WebJobs.Extensions.DurableTask/Correlation/ITelemetryActivator.cs]: I ended up removing this property.

File: src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs
Comment (bachuv) [reply src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs:73]: I like this idea! Existing apps that enabled distributed tracing wouldn't break and the OTel compatible implementation is opt-in.

File: src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs:26]: Any reason to not just call this what it is?
```suggestion
        private IAsyncDisposable telemetryModule;
```

File: src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs:73]: > We might as well keep the old stuff around but give a way to control what mode the customer uses.

@jviau just to clarify, are you suggesting we have this extra setting just for preview and then remove it for GA?

File: src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs:69]: Should we rename this method to something like `SetUpDeprecatedDistributedTracing()` to make it clear that this applies only to the old implementation?

**Review by cgillum — COMMENTED** [review]
A left a few comments/questions.

Also, I noticed that this code requires `APPINSIGHTS_INSTRUMENTATIONKEY`. However, shouldn't we also support [APPLICATIONINSIGHTS_CONNECTION_STRING](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings#applicationinsights_connection_string), which is recommended and also required for supporting sovereign clouds? I know this is technically outside the scope of this PR, but we should definitely do this work before releasing for public preview.

File: src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs
Comment (bachuv) [reply src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs]: Thanks for the suggestion! That makes sense. It's also used in `SetupTelemetryClient()`, but I just pass it in as a parameter now.

File: src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs
Comment (jviau) [reply src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs:73]: @cgillum I don't think we should remove the existing distributed tracing even after GA. At least not right away. We should give a deprecation window so customers that already use it have a chance to migrate over. We may want to track this work, but we should probably give an in-portal alert for customers to switch over to new distributed tracing, and give a timeline that it will be removed at a certain date (we can decide on that date).

**Comment by jviau** [comment]
> Also, I noticed that this code requires APPINSIGHTS_INSTRUMENTATIONKEY. However, shouldn't we also support [APPLICATIONINSIGHTS_CONNECTION_STRING](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings#applicationinsights_connection_string), which is recommended and also required for supporting sovereign clouds? I know this is technically outside the scope of this PR, but we should definitely do this work before releasing for public preview.

Better yet, we can just DI the existing `TelemetryClient` or `IOptions<TelemetryConfiguration>` and not worry about building it ourselves at all. Either of these types works for import, I tested it with the SmokeTest v3 sample.

File: src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs:73]: Okay, but we probably need @lilyjma to weigh in on this decision.

File: src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs:82]: This is a preview feature that we never intend to GA so V1 feels misleading. Deprecated feels like the right term to me, even before we GA the new experience.

Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs:82]: But just to argue with myself - I suppose another way to look at it is that this has been our only option for so long that we might as well treat it as if it was GA. From that perspective, I suppose it's fine to treat these as V1 and V2.

**Comment by cgillum** [comment]
> > Also, I noticed that this code requires APPINSIGHTS_INSTRUMENTATIONKEY. However, shouldn't we also support [APPLICATIONINSIGHTS_CONNECTION_STRING](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings#applicationinsights_connection_string), which is recommended and also required for supporting sovereign clouds? I know this is technically outside the scope of this PR, but we should definitely do this work before releasing for public preview.
> 
> Better yet, we can just DI the existing `TelemetryClient` or `IOptions<TelemetryConfiguration>` and not worry about building it ourselves at all. Either of these types works for import, I tested it with the SmokeTest v3 sample.

I'm fine with either approach. @bachuv can you look into this?

**Comment by bachuv** [comment]
> > > Also, I noticed that this code requires APPINSIGHTS_INSTRUMENTATIONKEY. However, shouldn't we also support [APPLICATIONINSIGHTS_CONNECTION_STRING](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings#applicationinsights_connection_string), which is recommended and also required for supporting sovereign clouds? I know this is technically outside the scope of this PR, but we should definitely do this work before releasing for public preview.
> > 
> > 
> > Better yet, we can just DI the existing `TelemetryClient` or `IOptions<TelemetryConfiguration>` and not worry about building it ourselves at all. Either of these types works for import, I tested it with the SmokeTest v3 sample.
> 
> I'm fine with either approach. @bachuv can you look into this?

@cgillum Yes, of course! I'll look into it for public preview.

File: src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs
Comment (jviau) [reply src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs:82]: > I suppose another way to look at it is that this has been our only option for so long that we might as well treat it as if it was GA

This is how I feel about it. Unless we find that leaving it in causes actual issues, then we might as well leave it be for existing users.

File: src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs
Comment (lilyjma) [reply src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs:73]: Thanks for looping me in. 

I've discussed with Varshi. Since replacing the old implementation would break current users, Jacob's suggestion sounds reasonable. Also, because the old implementation has been in preview for >1 year, we'll need to treat it as GA when retiring it, _assuming_ that it has >1k customers. Per Azure Retirement process, we'll need to give 18 month notice to retire the old implementation when there's a GA replacement (we could also replace with preview but that would require CTO approval). If we have <1k customers on the old implementation, the process would be easier. So the first step is to get the number of customers using the old implementation.

File: src/WebJobs.Extensions.DurableTask/Options/DurableDistributedTracingVersion.cs
Comment (jviau) [inline src/WebJobs.Extensions.DurableTask/Options/DurableDistributedTracingVersion.cs:15]: We should probably add a `None` value, just to set us up to remove the explicit enabled property on the options (or at least tweak it to interact with this enum, and in a future PR we can do this).

Make sure to then also update the `TelemetryActivator` to no-op when `None` is set.

``` CSharp
None = 0,
V1 = 1,
V2 = 2,
```

File: src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs
Comment (jviau) [inline src/WebJobs.Extensions.DurableTask/Options/TraceOptions.cs:73]: What do you think of naming the property `Version`?
```suggestion
        public DurableDistributedTracingVersion Version { get; set; } = DurableDistributedTracingVersion.V1;
```

**Review by jviau — APPROVED** [review]

**Comment by bachuv** [comment]
Thanks for approving, @jviau! I'll merge these changes after https://github.com/Azure/durabletask/pull/648 is merged since this PR is dependent on https://github.com/Azure/durabletask/pull/648.

@cgillum, can you take a look at the recent changes?

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:20]: Is this change required?

**Comment by cgillum** [comment]
> @cgillum Yes, of course! I'll look into it for public preview.

If you don't want to do Jacob's proposal for private preview, can we at least add support for `APPLICATIONINSIGHTS_CONNECTION_STRING` in the private preview? I feel like it would be a mistake to release any preview without support for both configuration options, especially `APPLICATIONINSIGHTS_CONNECTION_STRING` which is the only one we recommend.

**Comment by bachuv** [comment]
@cgillum, that makes sense! I added support for `APPLICATIONINSIGHTS_CONNECTION_STRING`.

**Review by cgillum — APPROVED** [review]
Looks good - thanks!

File: src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs:56]: I'm a bit concerned about this pattern, as I understand that calling `GetAwaiter()` is generally anti-pattern. Do we use disposing pattern (`.GetAwaiter().GetResult`) anywhere else in the codebase?

**Review by davidmrdavid — COMMENTED** [review]
Just a small question

File: src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs
Comment (bachuv) [reply src/WebJobs.Extensions.DurableTask/Correlation/TelemetryActivator.cs:56]: Yes, I see it used in DurableTaskExtension.cs ([example](https://github.com/Azure/azure-functions-durable-extension/blob/18d85bd8c0d41d5936cac46054e3839066ef65a7/src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs#L559))  and DurableTaskListener.cs ([example](https://github.com/Azure/azure-functions-durable-extension/blob/18d85bd8c0d41d5936cac46054e3839066ef65a7/src/WebJobs.Extensions.DurableTask/Listener/DurableTaskListener.cs#L88)).

**Review by davidmrdavid — APPROVED** [review]

**Pull request merged**
